utils/index.d.ts

```diff
@@ -1950,3 +1950,95 @@ export declare namespace randomUtils {
+    commandArgs?: unknown[];
+}
+
+/**
+ * Creates a generic element.
+ *
+ * @param options - options for the generic item
+ *
+ * If `options.commandArgs` is not set, then it will be set to the item itself.
+*/
+export declare function createGenericElement(options: GenericElementOptions): TreeElementBase;
+
+export declare interface TreeElementStateModel {
+    /**
+     * Apply a temporary description to the tree item
+     */
+    temporaryDescription?: string;
+    /**
+     * Set the tree item icon to a spinner
+     */
+    spinner?: boolean;
+    /**
+     * Temporary children to be displayed
+     */
+    temporaryChildren?: TreeElementBase[];
+}
+
+/**
+ * Wraps tree elements in a state model that can be used to temporarily override tree element behavior like the description or children.
+ */
+export declare class TreeElementStateManager<TElement extends TreeElementWithId = TreeElementWithId> implements Disposable {
+    /**
+     * Temporarily override the description of the given element while the callback is running.
+     *
+     * @param id - id of the element to temporarily override the tree item description of
+     * @param description - description to temporarily show
+     * @param callback - callback to run while the temporary description is shown
+     * @param dontRefreshOnRemove - If true, the tree item won't be refreshed after the temporary description is removed.
+     * Useful when the tree item is being deleted. This prevents any time between the tree item description being cleared
+     * and the tree item being removed from the tree.
+     */
+    runWithTemporaryDescription<T = void>(id: string, description: string, callback: () => Promise<T>, dontRefreshOnRemove?: boolean): Promise<T>;
+
+    /**
+     * Shows a deleting state for the given element while the callback is running.
+     * Method calls `runWithTemporaryDescription` and passes options for deleting.
+     *
+     * @param id - id of the element to show a deleting state for
+     * @param callback - callback to run while the deleting state is shown
+     */
+    showDeleting(id: string, callback: () => Promise<void>): Promise<void>;
+
+    /**
+     * Adds a child to the given element while the callback is running. Typically used to show a "Creating..." child.
+     *
+     * @param id - id of the element to add a creating child to
+     * @param label - label of the child
+     * @param callback - callback to run while the creating child is shown
+     */
+    showCreatingChild<T = void>(id: string, label: string, callback: () => Promise<T>): Promise<T>;
+
+    /**
+     * Notify a resource that its children have changed.
+     * Calls the refresh callback with the given id.
+     * @param id - The id of the element for which the children have changed
+     */
+    notifyChildrenChanged(id: string): void;
+
+    /**
+     * Wraps an element in a state model that can be used to update the tree item. An element should only be wrapped once.
```

Comment:
> An element should only be wrapped once.

Is there a convenient way to throw an error if they try to double-wrap an object?

Reply:
We could set a property on the tree element object to mark it as wrapped and throw an error if we see that property.

Reply:
I implemented this in https://github.com/microsoft/vscode-azuretools/pull/1448, but I'm not sure we want to do it. Open to opinions.

Reply:
I don't feel strongly, but it might save us from a very hard-to-debug issue in the future.